File: genetic-packer/src/main/java/genetic/packer/evolution/generation/creator/OngoingGenerationCreator.java
package genetic.packer.evolution.generation.creator;

import java.util.function.BiFunction;

import com.google.common.collect.Iterables;
import genetic.api.individual.Individual;
import genetic.api.individual.Organism;
import genetic.api.individual.impl.RatedIndividual;
import genetic.packer.evolution.elitism.ElitistPicker;
import genetic.packer.evolution.generation.GeneticContext;
import genetic.packer.evolution.generation.dto.Generation;
import genetic.packer.evolution.mutation.Mutator;
import genetic.recombinators.Recombinator;
import genetic.selectors.PairSelector;
import javafx.scene.shape.Box;
import javaslang.collection.List;
import javaslang.collection.Seq;
import javaslang.collection.Traversable;
import javaslang.control.Option;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

/**
 * @author piotr.larysz
 */
@Component
public class OngoingGenerationCreator implements BiFunction<Generation<Double, Box>, GeneticContext, Traversable<Individual<Double, Box>>> {

    @Autowired
    private ElitistPicker elitistPicker; //todo feature toggle for cloning?

    @Autowired
    private PairSelector.Creator<Double> pairSelectorCreator;

    @Autowired
    private Recombinator<Organism<Box>, Individual<Double, Box>> singleChildRecombinator;

    @Autowired
    private Recombinator<Organism<Box>, Individual<Double, Box>> twoChildrenRecombinator;

    @Autowired
    private Mutator mutator;

    @Override
    public Traversable<Individual<Double, Box>> apply(Generation<Double, Box> generation, GeneticContext evolutionContext) {

        final Seq<Individual<Double, Box>> elites = this.elitistPicker.pick(generation.getRatedIndividuals(), evolutionContext.getNumberOfEliteIndividuals()); //fixme remove sorting inside

        final int noIndividualsToGenerate = (generation.getRatedIndividuals().size() - elites.size()) / 2;
        final boolean numberOfNewIndividualsIsOdd = (generation.getRatedIndividuals().size() - noIndividualsToGenerate) % 2 != 0; //todo method (misc), check if proper

        final PairSelector<RatedIndividual<Double, Box>> pairSelector = this.pairSelectorCreator.from(generation.getRatedIndividuals());

        final Seq<Individual<Double, Box>> newIndividuals = pairSelector.select(size -> noIndividualsToGenerate)
                .flatMap(this.twoChildrenRecombinator)
                .appendAll(this.createSingleForEven(pairSelector, numberOfNewIndividualsIsOdd))
                .peek(individual ->  this.mutator.accept(individual, evolutionContext.getEmbryo().getBounds()));

        return List.<Individual<Double, Box>> empty().appendAll(newIndividuals).appendAll(elites);
    }

    private Iterable<Individual<Double,Box>> createSingleForEven(PairSelector<RatedIndividual<Double, Box>> pairSelector, boolean numberOfNewIndividualsIsOdd) {
        if(numberOfNewIndividualsIsOdd) {
            return Option.of(pairSelector.select(size -> 1))
                    .map(Iterables::getOnlyElement)
                    .map(singleChildRecombinator)
                    .map(Iterables::getOnlyElement);
        }
        return Option.none();
    }
}

File: genetic-generations/src/main/java/genetic/generations/creator/OngoingGenerationCreator.kt
package genetic.generations.creator

import genetic.api.generation.Generation
import genetic.api.generation.GenerationContext
import genetic.api.individual.Individual
import genetic.api.individual.Organism
import genetic.api.individual.impl.RatedIndividual
import genetic.api.mutation.Mutator
import genetic.evaluator.elitism.ElitistPicker
import genetic.recombinators.Recombinator
import genetic.selectors.PairSelector

class OngoingGenerationCreator<T, P>( //todo feature toggle for cloning????
    private val elitistPicker: ElitistPicker,
    private val pairSelectorFactory: PairSelector.Factory<RatedIndividual<P>>,
    private val singleChildRecombinator: Recombinator<Organism<P>, Individual<P>>,
    private val twoChildrenRecombinator: Recombinator<Organism<P>, Individual<P>>,
    private val mutator: Mutator<T, P>
) : (Generation<P>, GenerationContext<T>) -> Sequence<Individual<P>> {

  override fun invoke(
      generation: Generation<P>,
      evolutionContext: GenerationContext<T>
  ) = with(pairSelectorFactory(generation.ratedIndividuals)) {
    this.select()
        .take((generation.ratedIndividuals.size - evolutionContext.numberOfEliteIndividuals) / 2)
        .flatMap(twoChildrenRecombinator)
        .plus(
            createSingleForEven(
                pairSelector = this,
                numberOfNewIndividualsIsOdd = evolutionContext.numberOfEliteIndividuals % 2 != 0
            )
        )
        //.onEach { mutator(it, evolutionContext.embryo) } FIXME cast issue
        .plus(
            elitistPicker.pick(generation.ratedIndividuals)
                .take(evolutionContext.numberOfEliteIndividuals)
        )
  }

  private fun createSingleForEven(
      pairSelector: PairSelector<RatedIndividual<P>>,
      numberOfNewIndividualsIsOdd: Boolean
  ) = if (numberOfNewIndividualsIsOdd) {
    pairSelector.select()
        .take(1)
        .single()
        .let(singleChildRecombinator)
  } else emptySequence()
}